### PR TITLE
Add TensorGroup requirement to Dataset API

### PR DESCRIPTION
**Pull request by marcrasi**
This adds the TensorGroup requirement to the Dataset API, which gets the Dataset API working without deabstraction!

This allows us to delete UnknownShapeListAttribute and TypeListAttribute, so I deleted those!

## const evaluator bugfix

This exposed a bug in the const evaluator where it wasn't able to evaluate the `Element._typeList` and `Element._unknownShapeList` calls in the Dataset API.

Description of bug and fix:

There are two different ways of calculating substitution maps, depending on whether the function comes from a `function_ref` or `witness_method`. There is existing logic in `ConstExprFunctionState::computeCallResult` to handle this. For example,

Does a "normal" substitution map calculation because it's a `function_ref` (handled correctly by existing code):
```
// function_ref foo<A>(_:)
%5 = function_ref @$s5repro3fooySbxAA12ExampleProtoRzlF : $@convention(thin) <τ_0_0 where τ_0_0 : ExampleProto> (@in_guaranteed τ_0_0) -> Bool // user: %6
%6 = apply %5<ExampleStruct>(%3) : $@convention(thin) <τ_0_0 where τ_0_0 : ExampleProto> (@in_guaranteed τ_0_0) -> Bool // user: %8
```

Does a special "witness method" substitution map calculation because it's a `witness_method` (handled correctly by existing code):
```
%10 = witness_method $T, #ExampleProto.foo!1 : <Self where Self : ExampleProto><I> (Self) -> (I) -> Int : $@convention(witness_method: ExampleProto) <τ_0_0 where τ_0_0 : ExampleProto><τ_1_0> (@in_guaranteed τ_1_0, @in_guaranteed τ_0_0) -> Int // user: %11
%11 = apply %10<T, String>(%8, %0) : $@convention(witness_method: ExampleProto) <τ_0_0 where τ_0_0 : ExampleProto><τ_1_0> (@in_guaranteed τ_1_0, @in_guaranteed τ_0_0) -> Int // user: %15
```

However, this PR exposes a new situation that is handled incorrectly:

```
// function_ref protocol witness for static TensorGroup._typeList.getter in conformance Tensor<A>
%47 = function_ref @$s10TensorFlow0A0VyxGAA0A5GroupA2aEP9_typeListSayAA0A8DataTypeVGvgZTW : $@convention(witness_method: TensorGroup) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@thick Tensor<τ_0_0>.Type) -> @owned Array<TensorDataType> // user: %48
%48 = apply %47<Float>(%46) : $@convention(witness_method: TensorGroup) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@thick Tensor<τ_0_0>.Type) -> @owned Array<TensorDataType> // users: %52, %63
```

This last case confuses the existing code, because it sees that the function is `@convention(witness_method)` and tries to do the special "witness method" substitution map calculation. This is wrong. It should do the normal substitution map calculation because the function comes from a `function_ref`.

I fix this by keeping track of whether the function comes from a `function_ref` or a `witness_method`, and then using that piece of information to determine how to calculate the substitution map.

### CODE REVIEW

File: stdlib/public/TensorFlow/Dataset.swift [hunk 1]
```diff
@@ -182,23 +182,25 @@ extension DatasetIterator : IteratorProtocol {
...
 // FIXME: SR-8599 blocks heterogeneous scalar types.
-@inlinable @inline(__always)
-public func zip<T>(
-  _ dataset1: Dataset<T>, _ dataset2: Dataset<T>
-) -> Dataset<(T, T)> {
-  return #tfop("ZipDataset", [dataset1, dataset2],
-               output_types$typeList: (T, T).self,
-               output_shapes$unknownShapeList: (T, T).self)
-}
+// FIXME: Tuples can't conform to TensorGroup, so we will have to define a
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Dataset.swift]: You can define a `TensorPair` struct for this. Tuples datasets are extremely common so we should support it.

File: lib/ParseSIL/ParseSIL.cpp [hunk 2]
```diff
@@ -1019,7 +1019,34 @@ static bool parseSymbolicValue(SymbolicValue &value, SILParser &SP,
     SILLocation funcLoc = RegularLocation(P.getEndOfPreviousLoc());
     if (SP.parseSILFunctionRef(funcLoc, func))
       return true;
-    value = SymbolicValue::getFunction(func);
+
+    Identifier subConventionId;
+    FunctionSubConvention subConvention;
+    if (!P.Tok.is(tok::l_paren)) {
```
Comment (rxwei) [inline lib/ParseSIL/ParseSIL.cpp]: Please use `isNot` instead of inverting `is`.

File: lib/ParseSIL/ParseSIL.cpp [hunk 3]
```diff
@@ -1019,7 +1019,34 @@ static bool parseSymbolicValue(SymbolicValue &value, SILParser &SP,
...
+      P.diagnose(P.Tok, diag::sil_const_expected_fn_sub_conv);
+      return true;
+    }
+    P.consumeToken();
+    if (!P.Tok.is(tok::identifier)) {
```
Comment (rxwei) [inline lib/ParseSIL/ParseSIL.cpp]: Same here.

File: lib/ParseSIL/ParseSIL.cpp [hunk 4]
```diff
@@ -1019,7 +1019,34 @@ static bool parseSymbolicValue(SymbolicValue &value, SILParser &SP,
...
+    P.consumeIdentifier(&subConventionId);
+    if (subConventionId.str() == "N") {
+      subConvention = FunctionSubConvention::Normal;
+    } else if (subConventionId.str() == "W") {
+      subConvention = FunctionSubConvention::Witness;
+    } else {
+      P.diagnose(P.Tok, diag::sil_const_expected_fn_sub_conv);
+      return true;
+    }
+    if (!P.Tok.is(tok::r_paren)) {
```
Comment (rxwei) [inline lib/ParseSIL/ParseSIL.cpp]: Same here.

File: include/swift/SIL/SILConstants.h [hunk 5]
```diff
@@ -57,6 +57,19 @@ enum class UnknownReason {
   Trap,
 };
 
+/// Determines how to substitute the function's generic arguments when
+/// evaluating an apply of the function.
+enum class FunctionSubConvention {
```
Comment (rxwei) [inline include/swift/SIL/SILConstants.h]: Please name it something like `SubstitutionConvention`. "SubConvention" is very unclear to begin with -- subconvention sounds like a convention inside another convention.

**Review by rxwei — APPROVED** [review]
LGTM. This would break the tutorial. It should be updated accordingly. Speaking of that, the tutorial will need tensor tuples. Please add `TensorPair` to the TF module either in this patch or a follow-up patch.

Similar to #20188, this is a change to the public API. Please at least CC me on these PRs.

**Review by mhong — APPROVED** [review]
Great to see this patch simplified a lot of dataset infra code!

Discussed with Marc in person, and we also want to explore another option with the core team: after devirtualizing a function, can/should we change its convention from witness_method to sth else such as thin?

Marc will drive the discussion there. This patch can be landed to unblock our short term work.

File: lib/ParseSIL/ParseSIL.cpp [hunk 6]
```diff
@@ -1019,7 +1019,34 @@ static bool parseSymbolicValue(SymbolicValue &value, SILParser &SP,
     SILLocation funcLoc = RegularLocation(P.getEndOfPreviousLoc());
     if (SP.parseSILFunctionRef(funcLoc, func))
       return true;
-    value = SymbolicValue::getFunction(func);
+
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/ParseSIL/ParseSIL.cpp]: This is inside a tensorflow-specific function.

File: test/TensorFlow/graph_op_inst.sil [hunk 7]
```diff
@@ -15,9 +15,10 @@ bb0:
   %2 = graph_op "tf.Dummy"() {float1: f64 3.14, float2: f32 -3.14} : $Tensor<Float>
   %3 = graph_op "tf.Dummy"() {string1: "hello", string2: "world"} : $Tensor<Float>
   %4 = graph_op "tf.Dummy"() {metatype1: $Float, metatype2: $Tensor<Float>} : $Tensor<Float>
-  %5 = graph_op "tf.Dummy"() {function1: @chained_op_test : $@convention(thin) (Tensor<Float>, Tensor<Float>) -> Tensor<Float>} : $Tensor<Float>
-  %6 = graph_op "tf.Dummy"() {aggregate: ((i8 1, i32 -2), (f32 -1.0, $Float))} : $Tensor<Float>
-  %7 = graph_op "tf.Smarty"() {array: [$Int: i32 -2, i32 97]} : $Tensor<Float>
+  %5 = graph_op "tf.Dummy"() {function1: @chained_op_test : $@convention(thin) (Tensor<Float>, Tensor<Float>) -> Tensor<Float> (N)} : $Tensor<Float>
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply test/TensorFlow/graph_op_inst.sil]: It's to test the round-trip print/parse for both enum values.

File: stdlib/public/TensorFlow/Dataset.swift (hunk 1 above)
Comment (marcrasi) [reply stdlib/public/TensorFlow/Dataset.swift]: done

File: lib/ParseSIL/ParseSIL.cpp (hunk 2 above)
Comment (marcrasi) [reply lib/ParseSIL/ParseSIL.cpp]: done

File: lib/ParseSIL/ParseSIL.cpp (hunk 3 above)
Comment (marcrasi) [reply lib/ParseSIL/ParseSIL.cpp]: done

File: lib/ParseSIL/ParseSIL.cpp (hunk 4 above)
Comment (marcrasi) [reply lib/ParseSIL/ParseSIL.cpp]: done

File: include/swift/SIL/SILConstants.h (hunk 5 above)
Comment (marcrasi) [reply include/swift/SIL/SILConstants.h]: done

**Comment by marcrasi** [comment]
@rxwei, I'll do the TensorPair as a separate PR because I have a few uncertainties about how it should work that we can discuss in a separate PR. Also, I'll do an update pass on the tutorial soon.

I'll drive the possible devirtualize change mentioned by @mhong as part of upstreaming the generic functions part of the constexpr patch (https://github.com/apple/swift/pull/19662), because it's very related.

File: lib/SILOptimizer/Mandatory/TFConstExpr.cpp [hunk 8]
```diff
@@ -780,6 +781,7 @@ SubstitutionMap getWitnessMethodSubstitutions(SILModule &Module, ApplySite AI,
 /// information about the error.
 llvm::Optional<SymbolicValue>
 ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
+
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply lib/SILOptimizer/Mandatory/TFConstExpr.cpp]: removed!

File: stdlib/public/TensorFlow/Dataset.swift (hunk 1 above)
Comment (marcrasi) [reply stdlib/public/TensorFlow/Dataset.swift]: oops, pasted "done" into the wrong box :)

Will actually do as a separate PR, as mentioned in my top-level comment.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow linux gpu

**Pull request merged**